crates/aptos-crypto/src/p256/p256_keys.rs

```diff
@@ -0,0 +1,270 @@
+    /// Deserialize a P256PrivateKey. This method will check for private key validity: i.e.,
+    /// correct key length.
+    fn try_from(bytes: &[u8]) -> std::result::Result<P256PrivateKey, CryptoMaterialError> {
+        // Note that the only requirement is that the size of the key is 32 bytes, something that
+        // is already checked during deserialization of p256::ecdsa::SigningKey
+        P256PrivateKey::from_bytes_unchecked(bytes)
+    }
+}
+
+impl Length for P256PrivateKey {
+    fn length(&self) -> usize {
+        Self::LENGTH
+    }
+}
+
+impl ValidCryptoMaterial for P256PrivateKey {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes().to_vec()
+    }
+}
+
+impl Genesis for P256PrivateKey {
+    fn genesis() -> Self {
+        let mut buf = [0u8; P256_PRIVATE_KEY_LENGTH];
+        buf[P256_PRIVATE_KEY_LENGTH - 1] = 1;
+        Self::try_from(buf.as_ref()).unwrap()
+    }
+}
+
+//////////////////////
+// PublicKey Traits //
+//////////////////////
+
+// Implementing From<&PrivateKey<...>> allows to derive a public key in a more elegant fashion
+impl From<&P256PrivateKey> for P256PublicKey {
+    fn from(private_key: &P256PrivateKey) -> Self {
+        let secret = &private_key.0;
+        let public: p256::ecdsa::VerifyingKey = secret.into();
+        P256PublicKey(public)
+    }
+}
+
+// We deduce PublicKey from this
+impl PublicKey for P256PublicKey {
+    type PrivateKeyMaterial = P256PrivateKey;
+}
+
+impl std::hash::Hash for P256PublicKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let encoded_pubkey = self.to_bytes();
+        state.write(&encoded_pubkey);
+    }
+}
+
+// Those are required by the implementation of hash above
+impl PartialEq for P256PublicKey {
+    fn eq(&self, other: &P256PublicKey) -> bool {
+        self.to_bytes() == other.to_bytes()
+    }
+}
+
+impl Eq for P256PublicKey {}
+
+// We deduce VerifyingKey from pointing to the signature material
+// we get the ability to do `pubkey.validate(msg, signature)`
+impl VerifyingKey for P256PublicKey {
+    type SignatureMaterial = P256Signature;
+    type SigningKeyMaterial = P256PrivateKey;
+}
+
+impl fmt::Display for P256PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", hex::encode(self.0.to_sec1_bytes()))
+    }
+}
+
+impl fmt::Debug for P256PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "P256PublicKey({})", self)
+    }
+}
+
+impl TryFrom<&[u8]> for P256PublicKey {
+    type Error = CryptoMaterialError;
+
+    /// Deserialize a P256PublicKey.
+    fn try_from(bytes: &[u8]) -> std::result::Result<P256PublicKey, CryptoMaterialError> {
+        P256PublicKey::from_bytes_unchecked(bytes)
+    }
+}
+
+impl Length for P256PublicKey {
+    fn length(&self) -> usize {
+        P256_PUBLIC_KEY_LENGTH
+    }
+}
+
+impl ValidCryptoMaterial for P256PublicKey {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_sec1_bytes().to_vec()
```

Comment:
Should probably just be `self.to_bytes()` which in turn calls `to_sec1_bytes()`, for easier, single-place updates to the code.

Reply:
I don't follow - `self.to_bytes()` is what is being implemented in this function

Reply:
Yes, but here you are implementing `ValidCryptoMaterial::to_bytes`.

You already have a `P256PublicKey::to_bytes`, which I'm trying to suggest should be called here:

```
impl P256PublicKey {
    /// Serialize a P256PublicKey.
    pub fn to_bytes(&self) -> [u8; P256_PUBLIC_KEY_LENGTH] {
        // The RustCrypto P256 `to_sec1_bytes` call here should never return an array of the wrong length and cause a panic
        (*self.0.to_sec1_bytes()).try_into().unwrap()
    }
```

You can probably call it via `P256PublicKey::to_bytes(&self)`.

Otherwise, you have two places in which you call `to_sec1_bytes`, and when you change one place you might forget the other.

Reply:
You are correct - changed the line to `self.to_bytes().to_vec()`